src/app/binding/BindingTable.cpp

```diff
@@ -0,0 +1,237 @@
+                                             GetFabricInfoFromIndex(peerFabricIndex) };
+        OperationalDeviceProxy * device  = mPeerDevices.CreateObject(initParams, entry->mPeerId);
+        if (device == nullptr)
+        {
+            mBindingEntries.ReleaseObject(entry);
+            return CHIP_ERROR_NO_MEMORY;
+        }
+        return TryConnectDevice(device);
+    }
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR BindingTable::Remove(EndpointId sourceEndpoint, NodeId peerNodeId, EndpointId peerEndpointId,
+                                FabricIndex peerFabricIndex, ClusterId clusterId)
+{
+    BindingTableEntry * removeEntry = nullptr;
+    int peerNumBindings             = 0;
+    PeerId peer                     = BuildPeerId(peerNodeId, peerFabricIndex);
+    mBindingEntries.ForEachActiveObject(
+        [&removeEntry, &peerNumBindings, sourceEndpoint, peer, peerEndpointId, clusterId](BindingTableEntry * entry) {
+            if (entry->mSourceEndpoint == sourceEndpoint && entry->mPeerId == peer && entry->mPeerEndpointId == peerEndpointId &&
+                entry->mClusterId == clusterId)
+            {
+                removeEntry = entry;
+            }
+            if (entry->mPeerId == peer)
+            {
+                peerNumBindings++;
+            }
+            return true;
+        });
+    if (removeEntry == nullptr)
+    {
+        return CHIP_ERROR_KEY_NOT_FOUND;
+    }
+    mBindingEntries.ReleaseObject(removeEntry);
+    // Last binding with the peer, close the connection.
+    if (peerNumBindings == 1)
+    {
+        OperationalDeviceProxy * device = FindPeerDevice(peer);
+        if (device)
+        {
+            mPeerDevices.ReleaseObject(device);
+        }
+    }
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR BindingTable::TryConnectDevice(OperationalDeviceProxy * device)
+{
+    if (device == nullptr)
+    {
+        return CHIP_ERROR_INVALID_ARGUMENT;
+    }
+    if (device->IsConnected())
+    {
+        return CHIP_ERROR_INCORRECT_STATE;
+    }
+    if (device->IsConnecting())
+    {
+        return CHIP_NO_ERROR;
+    }
+    Dnssd::Resolver::Instance().SetResolverDelegate(&mResolveDelegate);
+    CHIP_ERROR ret = Dnssd::Resolver::Instance().ResolveNodeId(device->GetPeerId(), Inet::IPAddressType::kAny);
+    return ret;
+}
+
+BindingTableEntry * BindingTable::FindBindingEntry(NodeId node, FabricIndex fabricIndex)
+{
+    return FindBindingEntry(BuildPeerId(node, fabricIndex));
+}
+
+BindingTableEntry * BindingTable::FindBindingEntry(PeerId peer)
+{
+    BindingTableEntry * foundEntry = nullptr;
+    mBindingEntries.ForEachActiveObject([&foundEntry, peer](BindingTableEntry * entry) {
+        if (entry->mPeerId == peer)
+        {
+            foundEntry = entry;
+            return false;
+        }
+        return true;
+    });
+    return foundEntry;
+}
+
+BindingTableEntry * BindingTable::FindBindingEntry(EndpointId endpoint, ClusterId cluster)
```

Comment:
Why is this finding an entry, singular?  There can be multiple bindings for the same source endpoint and cluster.  When would this method be used and how would it work in that situation?

Reply:
Removed